### ISSUE TITLE
Grammar incompatible with lalr

### ISSUE THREAD

**Issue by Inky-developer** [lark-parser/lark#591]
I am writing a relatively simple grammar which gets parsed with lalr. After some changes it did not work anymore. I have reduced the erroneous grammar as in the following code snippet and it seems to function perfectly when the earley parser is used. All the grammar should do is match some content (which I am not interested in so I do not need to do anything with it) which is between two curly brackets. Because the content may contain curly brackets itself the grammar must check for balanced brackets. Here is minimal code:

```python
from lark import Lark

grammar = Lark(r"""
start: element ("," element)*
element: /[a-zA-Z]+/ "=" value

value: /[a-zA-Z0-9_]+/ | nbt

nbt: "{" nbt* "}"
   | /[^\{\}]+/
""", parser="lalr")

print(grammar.parse(r"player=me,nbt={data=[1,2,3,4]},x=10").pretty())
```

When this code is run the following error will be shown:

> lark.exceptions.UnexpectedToken: Unexpected token Token(__ANON_2, ',x=10') at line 1, column 31.
> Expected one of: 
> 	* $END
> 	* COMMA
> 

With the earley parser everything works out as intended. Are these two parsers supposed to behave differently on the same grammar? I thought the only difference is that some grammars do not work with lalr which do work with earley.

**Comment by erezsh** [comment]
>  I thought the only difference is that some grammars do not work with lalr which do work with earley.

Yes... and your grammar is one of them.

My guess is that your definition of nbt is too relaxed. You should try to either only match what you really need to, or include the curly brackets in the token to "segment" it.

**Comment by Inky-developer** [comment]
Thank you for your quick response. 

> You should try to either only match what you really need to, or include the curly brackets in the token to "segment" it.

If possible I would like to avoid writing a grammar to match nbt because it is really not important what it contains. What exactly do you mean with including curly brackets in the token?

**Comment by erezsh** [comment]
Right now nbt is a rule, which means that its parts are matched by the lexer, and then combined in the parser. You can make it into a terminal, which will match everything in the lexer, and avoid some ambiguity in the parser. Something like:

    NBT: "{" /[^\{\}]+/ "}"

Of course, it won't count the curlies to make sure they are matched. But you can do that later on.

**Comment by Inky-developer** [comment]
Wouldn't this token fail to match after the first inner bracket?

I think I will just go with the Earley parser.

Thank you for your help, this library is truly amazing!

**Comment by MegaIng** [comment]
This is once again this: https://gitter.im/lark-parser/Lobby?at=5ea8a3755cd4fe50a3d8ede9

Check my explanation here: https://gitter.im/lark-parser/Lobby?at=5ea8c694a9de3d01b1d709c2

TLDR: Don't create colliding tokens. The longer one will override the shorter one. (In you case `","` and `/[^\{\}]+/`)
